File: app/src/test/java/com/huhx0015/androidbooster/viewmodel/RecyclerViewModelTest.java
package com.huhx0015.androidbooster.viewmodel;

import com.huhx0015.androidbooster.application.AndroidApplication;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.mockito.junit.MockitoJUnitRunner;
import static org.mockito.Mockito.mock;

@RunWith(MockitoJUnitRunner.class)
public class RecyclerViewModelTest {

    /** CLASS VARIABLES ________________________________________________________________________ **/

    // MOCK VARIABLES
    @Mock
    private ApiRecyclerViewModel mViewModel;
    @Mock
    private AndroidApplication mApplication;

    // TEST VARIABLES
    private static final String TEST_ERROR_TEXT = "An error occurred.";
    private static final String TEST_ERROR_TEXT_ALTERNATE = "An unknown error occurred.";

    /** INIT METHODS ___________________________________________________________________________ **/

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        mApplication = mock(AndroidApplication.class);

        mViewModel = new ApiRecyclerViewModel(mApplication);
        mViewModel.setErrorText(TEST_ERROR_TEXT);
    }

    /** TEST METHODS ___________________________________________________________________________ **/

    @Test
    public void testGetErrorVisibility() {
        Assert.assertFalse(mViewModel.getErrorVisible());
    }

    @Test
    public void testGetProgressBarVisibility() {
        Assert.assertFalse(mViewModel.getProgressBarVisible());
    }

    @Test
    public void testGetRecyclerViewVisibility() {
        Assert.assertFalse(mViewModel.getRecyclerViewVisible());
    }

    @Test
    public void testGetErrorText() {
        Assert.assertEquals(TEST_ERROR_TEXT, mViewModel.getErrorText());
    }

    @Test
    public void testSetErrorVisibility() {
        mViewModel.setErrorVisible(true);
        Assert.assertTrue(mViewModel.getErrorVisible());
    }

    @Test
    public void testSetProgressBarVisibility() {
        mViewModel.setProgressBarVisible(true);
        Assert.assertTrue(mViewModel.getProgressBarVisible());
    }

    @Test
    public void testSetRecyclerViewVisibility() {
        mViewModel.setRecyclerViewVisible(true);
        Assert.assertTrue(mViewModel.getRecyclerViewVisible());
    }

    @Test
    public void testSetErrorText() {
        mViewModel.setErrorText(TEST_ERROR_TEXT_ALTERNATE);
        Assert.assertEquals(TEST_ERROR_TEXT_ALTERNATE, mViewModel.getErrorText());
    }
}

File: app/src/test/java/com/huhx0015/androidbooster/viewmodel/RecyclerViewModelTest.kt
package com.huhx0015.androidbooster.viewmodel

import android.app.Application
import androidx.databinding.PropertyChangeRegistry
import io.reactivex.rxjava3.disposables.CompositeDisposable
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.MockitoAnnotations
import org.mockito.junit.MockitoJUnitRunner
import org.mockito.Mockito.mock

@RunWith(MockitoJUnitRunner::class)
class RecyclerViewModelTest {

    /** CONSTANTS _______________________________________________________________________________ */

    companion object {
        private const val TEST_ERROR_TEXT = "An error occurred."
        private const val TEST_ERROR_TEXT_ALTERNATE = "An unknown error occurred."
    }

    /** CLASS VARIABLES ________________________________________________________________________  */

    // MOCK VARIABLES
    @Mock
    lateinit var viewModel: ApiRecyclerViewModel
    @Mock
    lateinit var application: Application

    /** INIT METHODS ___________________________________________________________________________  */

    @Before
    fun setUp() {
        MockitoAnnotations.initMocks(this)
        application = mock(Application::class.java)

        viewModel = ApiRecyclerViewModel(application)
        // Manually initialize Hilt-injected dependencies for testing
        viewModel.disposable = CompositeDisposable()
        viewModel.registry = PropertyChangeRegistry()
        viewModel.errorText = TEST_ERROR_TEXT
    }

    /** TEST METHODS ___________________________________________________________________________  */

    @Test
    fun testGetErrorVisibility() {
        Assert.assertFalse(viewModel.errorVisible)
    }

    @Test
    fun testGetProgressBarVisibility() {
        Assert.assertFalse(viewModel.progressBarVisible)
    }

    @Test
    fun testGetRecyclerViewVisibility() {
        Assert.assertFalse(viewModel.recyclerViewVisible)
    }

    @Test
    fun testGetErrorText() {
        Assert.assertEquals(TEST_ERROR_TEXT, viewModel.errorText)
    }

    @Test
    fun testSetErrorVisibility() {
        viewModel.errorVisible = true
        Assert.assertTrue(viewModel.errorVisible)
    }

    @Test
    fun testSetProgressBarVisibility() {
        viewModel.progressBarVisible = true
        Assert.assertTrue(viewModel.progressBarVisible)
    }

    @Test
    fun testSetRecyclerViewVisibility() {
        viewModel.recyclerViewVisible = true
        Assert.assertTrue(viewModel.recyclerViewVisible)
    }

    @Test
    fun testSetErrorText() {
        viewModel.errorText = TEST_ERROR_TEXT_ALTERNATE
        Assert.assertEquals(TEST_ERROR_TEXT_ALTERNATE, viewModel.errorText)
    }
}